From f3ca71c9900fe860849b3163a6e2c1e765b291d9 Mon Sep 17 00:00:00 2001
From: Alexei Podtelezhnikov <apodtele@gmail.com>
Date: Mon, 7 Sep 2026 15:02:23 -0400
Subject: [PATCH] * src/cid/cidload.c (cid_read_subrs): Limit overlaps.

Overlaps between subroutine dictionaries can be useful but prone to
abuses. The arbitrary limit can be adjusted in the future. Fixes #1457.
---
 src/cid/cidload.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/cid/cidload.c b/src/cid/cidload.c
index 1a0e6b156..b9ab1b9e0 100644
--- a/src/cid/cidload.c
+++ b/src/cid/cidload.c
@@ -517,6 +517,7 @@
     CID_Subrs      subr;
     FT_UInt        max_offsets = 0;
     FT_ULong*      offsets = NULL;
+    FT_ULong       cum_len = 0;
     PSAux_Service  psaux = (PSAux_Service)face->psaux;
 
 
@@ -584,6 +585,14 @@
       /* now, compute the size of subrs charstrings, */
       /* allocate, and read them                     */
       data_len = offsets[num_subrs] - offsets[0];
+      cum_len += data_len;
+
+      if ( cum_len / 4 > stream->size )
+      {
+        FT_ERROR(( "cid_read_subrs: excessive overlaps\n" ));
+        error = FT_THROW( Invalid_File_Format );
+        goto Fail;
+      }
 
       if ( FT_QNEW_ARRAY( subr->code, num_subrs + 1 ) ||
            FT_QALLOC( subr->code[0], data_len )       )
-- 
GitLab

